src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/FormFileModelBinder.cs

```diff
@@ -36,66 +36,151 @@ public Task BindModelAsync(ModelBindingContext bindingContext)
         {
-            // If we're at the top level, then use the FieldName (paramter or property name).
-            // This handles the fact that there will be nothing in the ValueProviders for this parameter
-            // and so we'll do the right thing even though we 'fell-back' to the empty prefix.
-            var modelName = bindingContext.IsTopLevelObject
-                ? bindingContext.BinderModelName ?? bindingContext.FieldName
-                : bindingContext.ModelName;
 
-            object value;
+            ICollection<IFormFile> collection;
             if (bindingContext.ModelType == typeof(IFormFile))
             {
-                var postedFiles = await GetFormFilesAsync(modelName, bindingContext);
-                value = postedFiles.FirstOrDefault();
+                // Create an intermediate list. Will throw it away later.
+                collection = new List<IFormFile>();
             }
-            else if (typeof(IEnumerable<IFormFile>).IsAssignableFrom(bindingContext.ModelType))
+            else if (bindingContext.ModelType == typeof(IFormFile[]))
             {
-                var postedFiles = await GetFormFilesAsync(modelName, bindingContext);
-                value = ModelBindingHelper.ConvertValuesToCollectionType(bindingContext.ModelType, postedFiles);
+                if (bindingContext.ModelMetadata.IsReadOnly)
+                {
+                    // Can't change the length of an existing array or replace it.
+                    return null;
+                }
+
+                // Use a List<IFormFile> for now. Will create an array later.
+                collection = new List<IFormFile>();
             }
             else
             {
-                // This binder does not support the requested type.
-                Debug.Fail("We shouldn't be called without a matching type.");
-                return;
+                if (bindingContext.Model == null)
+                {
+                    if (bindingContext.ModelMetadata.IsReadOnly)
+                    {
+                        // Need a new collection instance but unable to assign it to the property.
+                        return null;
+                    }
+
+                    if (bindingContext.ModelType == typeof(IFormFileCollection))
+                    {
+                        // Special-case creating a custom IFormFileCollection if the property is settable and
+                        // currently null. Use a List<IFormFile> for now. Will create a FileCollection later.
+                        collection = new List<IFormFile>();
+                    }
+                    else
+                    {
+                        // Note this call may return null if the model type cannot be activated.
+                        collection = ModelBindingHelper.CreateCompatibleCollection<IFormFile>(
+                            bindingContext.ModelType,
+                            capacity: null);
```

Comment:
> I'd assume the header code is pretty similar without the `if` right?

Actually will have a similar layout except the `if` condition will be `bindingContext.ModelType == typeof(string)` because no collection, intermediate or otherwise, is needed in that case.

I'm going to try a few things out and see what feels best, keeping `GenericModelBinder` and friends in mind...